### PR TITLE
Use asymmetric average when using Malitsky & Pock linesearch

### PR DESCRIPTION
**Pull request by dapplegate**
Addresses a TODO in the Malitsky & Pock linesearch code.  The theory requires using an asymmetric average - including the primal starting point in the average (with a somewhat unusual weight), but not the dual starting point.  This changes the code to match the theory.

However, one caveat: on the one test instance I've run, academictimetablesmall with defaults except --method pdhg --step_size_policy=malitsky-pock, this change increases the number of iterations to solve from 1241 to 2041 (and kkt passes from 1.9e3 to 3.2e3).  I haven't run a larger benchmark.  I'm not sure this matters, but I found it interesting.

Note that supporting this required splitting the denominator in the average structure into primal and dual portions.  There were several places that tested whether the denominator was nonzero (before computing an average) - this changes them to require both denominators to be nonzero.  For the current code, this change is a no-op - they will be non-zero together, but it makes it clearer at the call site what's going on, rather than just testing one or the other.

### CODE REVIEW

**Comment by mlubin** [comment]
Is it worth adding unit tests for SolutionWeightedAverage?

**Comment by dapplegate** [comment]
We could.  That seems like mostly a separate issue - although this does change SolutionWeightedAverage, most of the usage is unchanged, so tested as much as it was before.

**Comment by mlubin** [comment]
> We could. That seems like mostly a separate issue - although this does change SolutionWeightedAverage, most of the usage is unchanged, so tested as much as it was before.

Fair enough.

@mateodd25 Could you take a quick look to confirm this matches up with the theory?

File: src/saddle_point.jl [hunk 1]
```diff
@@ -351,6 +391,11 @@ mutable struct RestartParameters
   """
   primal_weight_update_smoothing::Float64
 
+  """
+  Use an approximate localized duality gap for restart computations. The
```
Comment (mateodd25) [inline src/saddle_point.jl:395]: This seems to be part of a different PR.

File: src/saddle_point.jl [hunk 2]
```diff
@@ -618,12 +663,12 @@ accordingly.
    this struct will be reset to the restart point.
 - `current_primal_solution::AbstractVector`: If there is a restart then this
    vector might be set to the avg_primal_solution. However, it could remain
-   unchanged, if the algorithm thinks the averaged iterate is better than the
-   current iterate.
+   unchanged, if the algorithm thinks the current iterate is better than the
+   averaged iterate.
```
Comment (mateodd25) [inline src/saddle_point.jl:667]: Same comment as above. These changes should go on a separate PR.

**Review by mateodd25 — COMMENTED** [review]
The changes seem good. I recall implementing a solution with one single data structure, WeightedAverage, that we instantiated twice (for the primal and dual variables, respectively). Any reason why using a single data structure instead? I don't feel strongly about this; I am happy with the current structure.

File: src/saddle_point.jl (hunk 1 above)
Comment (dapplegate) [reply src/saddle_point.jl]: Moved to another PR

File: src/saddle_point.jl (hunk 2 above)
Comment (dapplegate) [reply src/saddle_point.jl]: Moved to another PR

**Comment by dapplegate** [comment]
> I recall implementing a solution with one single data structure, WeightedAverage, that we instantiated twice (for the primal and dual variables, respectively). Any reason why using a single data structure instead? I don't feel strongly about this; I am happy with the current structure.

I considered switching to separate primal/dual structures.  However, since all existing usage treated it as a single averaging object, leaving it that way and just changing the internals resulted in a smaller overall change, and I didn't feel strongly enough about having two simpler structures to make the larger change.

**Review by mateodd25 — APPROVED** [review]

**Review by mlubin — APPROVED** [review]
